### ISSUE TITLE
'ERROR TypeError field[type] is not iterable' how to fix?

### ISSUE THREAD

**Issue by Shutainzu-Geeto** [ngx-formly/ngx-formly#2818]
**Description**

Hello, 
We're not using formly the normal way so it's a bit harder for us to work around some of our issues, but the one we got here seems unrelated to our special environment but i'll try to explain since i'm not sure i can provide an example

We have a form with a custom formly chips component dispatching multiple mat-expansion-panel.
They contain mainly multiple toggle or select and are ordered in group of row.

Field's name,key... are auto generated based on the chips component selected.

While loading the app, the formly is generated base on our DB's data.
We don't have any validator yet as we're using default value for the time being.

The strange behavior here is the following: 
- While removing a component from the chips component, everything is working fine (despite this error showing up)
![Screenshot_20210427_163542](https://user-images.githubusercontent.com/4579829/116260010-9d2ec680-a776-11eb-902d-eef1e1014a6c.png)
- Adding a chips make it all bugged out and show multiple time the above error. 
It was working fine last week and we've check everything we could think about and can't figure out what's wrong. 
Everything seems to be good on our side but somehow what we can log and see doesn't match.
The model IS updated but as fo
This error is too abstract and doesn't help us figured out what's happening here.

Expected Behavior (working on init and delete)
![Screenshot_20210427_170009 (1)](https://user-images.githubusercontent.com/4579829/116265004-087a9780-a77b-11eb-82db-aa9de58b5c10.png)

Behavior while adding new component
![Screenshot_20210427_164258](https://user-images.githubusercontent.com/4579829/116261981-73769f00-a778-11eb-8f03-7d870944e9e9.png)

As you can see the new component is bugged and errors keep showing up
![Screenshot_20210427_170847](https://user-images.githubusercontent.com/4579829/116265277-45df2500-a77b-11eb-9a5f-c87bce01e4ef.png)

We've compare the before/after and everything is correct (can't show a complete stringified version as there are function in the object, sorry for that bad way of displaying the comparaison :upside_down_face: )
![Screenshot_20210427_172349 (1)](https://user-images.githubusercontent.com/4579829/116269979-fb5fa780-a77e-11eb-90b1-62d12edea62e.png)

**Your Environment**

"resolutions": {
    "webpack": "^5.0.0"
  },
  "dependencies": {
    "@angular/animations": "^11.2.8",
    "@angular/cdk": "^11.2.8",
    "@angular/common": "^11.2.8",
    "@angular/compiler": "^11.2.8",
    "@angular/core": "^11.2.8",
    "@angular/fire": "^6.1.4",
    "@angular/forms": "^11.2.8",
    "@angular/material": "^11.2.8",
    "@angular/platform-browser": "^11.2.8",
    "@angular/platform-browser-dynamic": "^11.2.8",
    "@angular/router": "^11.2.8",
    "@ngrx/component": "^11.1.0",
    "@ngrx/component-store": "^11.1.0",
    "@ngrx/data": "^11.1.0",
    "@ngrx/effects": "^11.1.0",
    "@ngrx/entity": "^11.1.0",
    "@ngrx/router-store": "^11.1.0",
    "@ngrx/store": "^11.1.0",
    "@ngrx/store-devtools": "^11.1.0",
    "@ngx-formly/core": "^5.10.15",
    "@ngx-formly/material": "^5.10.15",
    "angularx-qrcode": "^11.0.0",
    "axios": "^0.21.1",
    "camelcase": "^6.2.0",
    "clone-deep": "^4.0.1",
    "fast-deep-equal": "^3.1.3",
    "firebase": "^8.3.2",
    "firebase-admin": "^9.6.0",
    "firebase-functions": "^3.13.2",
    "glob": "^7.1.6",
    "immer": "^9.0.1",
    "koa": "^2.13.1",
    "koa-tree-router": "^0.7.0",
    "koa2-cors": "^2.0.6",
    "material-design-icons": "^3.0.1",
    "ngrx-store-freeze": "^0.2.4",
    "ngx-fancy-logger": "^1.0.2",
    "ngx-markdown": "^11.1.2",
    "ngx-mask": "^11.1.5",
    "nodemailer": "^6.5.0",
    "rxjs": "^6.6.7",
    "sharp": "^0.28.1",
    "stanza": "^12.15.0",
    "stripe": "^8.142.0",
    "swiper": "^6.5.5",
    "tslib": "^2.2.0",
    "uuid": "^8.3.2",
    "zone.js": "~0.11.4"
  },
  "devDependencies": {
    "@angular-architects/ddd": "^1.3.0",
    "@angular-devkit/build-angular": "^0.1102.8",
    "@angular-devkit/schematics": "^11.2.8",
    "@angular-eslint/eslint-plugin": "^2.0.2",
    "@angular-eslint/eslint-plugin-template": "^2.0.2",
    "@angular-eslint/template-parser": "^2.0.2",
    "@angular/cli": "^11.2.8",
    "@angular/compiler-cli": "^11.2.8",
    "@angular/language-service": "^11.2.8",
    "@angular/localize": "^11.2.8",
    "@ngrx/schematics": "^11.1.0",
    "@ngrx/store-devtools": "^11.1.0",
    "@ngtools/webpack": "^11.2.8",
    "@nrwl/angular": "^12.0.0",
    "@nrwl/cli": "^12.0.0",
    "@nrwl/cypress": "^12.0.0",
    "@nrwl/eslint-plugin-nx": "^12.0.0",
    "@nrwl/jest": "^12.0.0",
    "@nrwl/node": "^12.0.0",
    "@nrwl/schematics": "^8.12.11",
    "@nrwl/tao": "^12.0.0",
    "@nrwl/workspace": "^12.0.0",
    "@schematics/angular": "^11.2.8",
    "@schematics/package-update": "^0.7.5",
    "@types/fs-extra": "^9.0.9",
    "@types/glob": "^7.1.3",
    "@types/jest": "^26.0.22",
    "@types/koa": "^2.13.1",
    "@types/koa2-cors": "^2.0.1",
    "@types/node": "^14.14.37",
    "@types/nodemailer": "^6.4.1",
    "@types/sharp": "^0.28.0",
    "@types/swiper": "^5.4.2",
    "@types/uuid": "^8.3.0",
    "@typescript-eslint/eslint-plugin": "^4.21.0",
    "@typescript-eslint/parser": "^4.21.0",
    "copy-webpack-plugin": "^8.1.1",
    "cypress": "^7.0.1",
    "dotenv": "^8.2.0",
    "eslint": "^7.23.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-plugin-cypress": "^2.11.2",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-jsdoc": "^32.3.0",
    "eslint-plugin-prefer-arrow": "^1.2.3",
    "firebase-functions-test": "^0.2.3",
    "fs-extra": "^9.1.0",
    "handlebars": "^4.7.7",
    "jest": "^26.6.3",
    "jest-preset-angular": "^8.4.0",
    "ng-packagr": "^11.2.4",
    "npm-run-all": "^4.1.5",
    "prettier": "^2.2.1",
    "promise": "^8.1.0",
    "ts-jest": "^26.5.4",
    "ts-node": "^9.1.1",
    "typedoc": "^0.20.35",
    "typescript": "4.0.5",
    "webpack": "^5.31.0"
  }
 
  
  **Example**
  
  I'm not sure we can reproduce this as everything looks fine and the form + data is generated perfectly but not rendered as such... 
  I'll try to implement a stackblitz if we dont find anything but in the meantime we'd like to focus on that non-helping error :man_shrugging: 

Any help or suggestion would be appreciate

**Comment by aitboudad** [comment]
some fields are not processed by the builder, are you adding fields on the fly?
if so could you please re-trigger the build by changing the `fields` input reference, if not possible try to call: `options._buildForm()`

**Comment by Shutainzu-Geeto** [comment]
Yes fields are added on the fly!
I've tried using 
```typescript  
private checkField(){
  (this.options as any)._checkField({
    fieldGroup: this.fields,
    model: this.model,
    formControl: this.form,
    options: this.options,
  });
}
```
and 
  
```typescript 
 (<FormlyFormOptionsCache>this.options)._buildForm();
```
 
but doesn't seems to do anything, am i missing something?

**Comment by aitboudad** [comment]
> but doesn't seems to do anything, am i missing something?

nop, if you can provide a reproduction that would be better any way I've pushed a fix for the raised error but can't figure out the root issue in your code.

**Comment by Shutainzu-Geeto** [comment]
Yea i guess so too, it's kind of complex, i'm trying to make a stackblitz but i'm not sure i'll be able to reproduce it as i've said :/ i'll try tomorow

**Comment by aitboudad** [comment]
ok then, I'll postpone the release to Tomorrow :).

You may try by adding my fix to check if everything work as expected:
```patch
-   } else {
+   } else if (field[type]) {
     validators.push(...field[type]);
    }
```

**Comment by aitboudad** [comment]
```
"@ngx-formly/core": "^5.10.15",
"@ngx-formly/material": "^5.10.15",
```

Also, make sure you're using the latest version https://github.com/ngx-formly/ngx-formly/releases/tag/v5.10.18

**Comment by Shutainzu-Geeto** [comment]
I've tried to make a stackblitz to reproduce it, you can see the structure and all but while trying to reproduce it, i've made another error instead...
i'm still trying to fix it as of now and then i'll see if the error we got is replicated here!
https://stackblitz.com/edit/angular-formly-typeerror-fieldtype-is-not-iterable?file=src%2Fapp%2Fcomponent%2Fchips.ts

**Comment by Shutainzu-Geeto** [comment]
Now that's my error(that was a pain to implement :sweat:)!
The only différence here is that removing an input from my chips component is bugged, and the error message is **slightly** different!

```
ERROR TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
```
instead of 
```
ERROR TypeError: field[type] is not iterable (cannot read property undefined)
```
The difference is probably related with my chips error on delete tho, i'll just need to figure it out

**Comment by aitboudad** [comment]
Thanks for the reproduction, I'll check it soon also I think the way you've managed the chips with mat-expansion-panel can be simplified (not sure if you've tried using array type: https://formly.dev/examples/advanced/repeating-section-input).

**Comment by Shutainzu-Geeto** [comment]
Well... i just want to say WHAT THE FUCK????
we've refactored some part of our code(same for the stackblitz).
The stackblitz is working PERFECTLY FINE NOW! :upside_down_face: 
But our code isn't.... and it's basically the same thing except the stackblitz is simplified...

**Comment by aitboudad** [comment]
The error still there I think, do not update the `stackblitz` as I may need it for local reproduction.
![Selection_060](https://user-images.githubusercontent.com/1753742/116691488-62ad6f80-a9aa-11eb-88df-1f1d6f1615b8.png)

**Comment by Shutainzu-Geeto** [comment]
yea, i've changed it only to improve our change,, now that it exactly replicate our case i've frooze it

**Comment by Shutainzu-Geeto** [comment]
We've made some progress on our side
The error seems to be from the chips component, but i can't explain why it was working before then.

Inside formly-page.ts in myListValues function, we've removed the following line
```typescript
_this.fields = [..._this.fields];
```
It was causing an error that was duplicating the code -> duplicating the function -> *2 *2 *2 *2... every call.
It leaded us to that error:
![Screenshot_20210430_164454](https://user-images.githubusercontent.com/4579829/116711697-65708a80-a9d3-11eb-898b-13b43af02a99.png)

By searching the root of it, we found out that every occurence of the formcontrol's update in the chips stoped here
Line such as 
```typescript
this.formControl.setValue([...])
```
That are respectly in the add, selected, and remove function lead to that error.
I'm not exactly sure of why exactly...

**Comment by aitboudad** [comment]
I figure out the root cause of this issue, looking for fix a now.

> Edit: somehow it has worked for a few minute, and now it's not(nothing has changed in between) anymore that's our exact case/errors!

It happens to me too which a strange behavior (may be related to a bug in Angular), reloading the page is the solution to get back the error.

**Comment by Shutainzu-Geeto** [comment]
One of my colleague seems to have found a solution this week-end but a lot of the code has changed in between but now it's working fine with no error. i'm still trying to process what exactly was the reason for the error, so i'm not closing this issue yet as no clear answer/explanation has been provided

**Comment by Shutainzu-Geeto** [comment]
I'll try to describe what we did to fix it according to my colleague's explanation (i kind of understand but am not conviced yet as some behavior were too strange).

As i've said, we removed 
```typescript
_this.fields = [..._this.fields];
```
since it was causing duplication function.
We noticed another error in the chips this time. 
The error was strange to me so i cant explain exactly but things like this were change
```typescript
this.formControl.setValue([...this.formControl.value, event.option.viewValue]);
//to
`this.formControl.setValue([...(this.formControl?.value || []), event.option.viewValue]);`
```
this line was delete in ngOnDestroy()
```typescript
    this.onDestroy$.complete();
```

doesn't seems to change anything to me but whattever.

**Then the "ROOT" of it.**
From my colleague, the root of the problem seems to be that i'm generating/deleting on the fly. I then have some missmatch with my model. To fix it what we did is made our model know ALL of our component (every case in the chips has a config generated) and we just hide them with an ngIf. 
The downside of it is we had to remove the field's defaultValue as it would have force every chips component to be displayed in the model.

If i have some free time later i'll try to do a fixed stackblitz version of the issue in case it can help anyone. 

As Aitboudad said too, there might be some bug somewhere in angular or formly? i dont know but i dont think we fixed it, i just think we've just bypassed it. The errors thrown were not really helping to pinpoint the issue, and as it's not clearly identified yet it'll be troublesome to find a way to display better error message.

I'll let that issue open for now as i know you're working on it. i'll let you close it when you'll provide an explanation/fix. Don't mind closing it before i post my stackblitz fix, i'll just edit if that's the that case ;) (if we can edit a closed issue? :upside_down_face: )
